File: cw4.cpp
#include <iostream>
#include <vector>

/**
 * @brief Рекурсивно вычисляет сумму элементов массива (вектора).
 * * @param arr Ссылка на вектор целых чисел (передача по константной ссылке для эффективности).
 * @param index Текущий индекс, с которого начинается суммирование.
 * @return Сумма элементов вектора, начиная с текущего индекса.
 */
int sumArrayRecursive(const std::vector<int>& arr, int index) {
    
    // 1. Базовый случай (Условие завершения): 
    // Если индекс достиг конца вектора (его размера), возвращаем 0.
    if (index == arr.size()) {
        return 0; 
    }

    // 2. Рекурсивный шаг: 
    // Возвращаем текущий элемент (arr[index]) 
    // плюс рекурсивный вызов для следующего элемента (index + 1).
    return arr[index] + sumArrayRecursive(arr, index + 1);
}

int main() {
    // Пример использования
    std::vector<int> myArray = {10, 2, 5, 8};
    
    // Вызываем функцию, начиная с индекса 0
    int result = sumArrayRecursive(myArray, 0);

    std::cout << "Массив: [10, 2, 5, 8]" << std::endl;
    std::cout << "Сумма элементов (рекурсивно): " << result << std::endl;
    // Ожидаемый вывод: 25
    
    return 0;
}

File: cw4.py
def sum_array_recursive(arr):
    """
    Рекурсивно вычисляет сумму элементов массива (списка).

    :param arr: Список чисел.
    :return: Сумма элементов списка.
    """
    # 1. Базовый случай: Пустой массив
    if not arr:
        return 0
    
    # 2. Рекурсивный шаг: 
    # Возвращаем первый элемент + сумму остатка массива
    else:
        # arr[0] - первый элемент
        # arr[1:] - срез, содержащий все элементы, кроме первого
        return arr[0] + sum_array_recursive(arr[1:])



my_array = [10, 2, 5, 8]
result = sum_array_recursive(my_array)

print(f"Массив: {my_array}")
print(f"Сумма элементов (рекурсивно): {result}") 
